### PR TITLE
Top-level CMake

### LINKED ISSUE #468: [TASK] Merge CMakeLists into a single project
## Description

Currently, each of the components (framework, common, agent, controller) has a separate "top-level" CMakeLists.txt. That makes it unnecessarily difficult to build the whole - in practice, it has to be done with maptools.py. But using such a script bypasses the standard cmake integration in OpenWrt and in RDK-B entirely. In addition, the separate top-level CMakeLists.txt make it more difficult to set or change global options.

Therefore, there should be a single top-level CMakeLists.txt that builds all the components.

## Exit criteria

The full prplMesh stack can be built with a single top-level call to `cmake`.

## Detailed description

- A previous attempt is still available in the feature/cmake-top-level but that went too far.
- BML (and its dependencies) should still be exported; the rest doesn't have to be. See #1.

### PR DESCRIPTION
**Pull request by arnout**
For proper integration with build systems (OpenWrt and RDK-B) it is much
easier if we can build all of prplMesh in a single go with cmake,
without the need for maptools.py to build all components.

This PR moves the framework, common, controller and agent components to the top level.

Fixes #468.

### CODE REVIEW

**Review by tomereli — APPROVED** [review]
So far, so good. But this means that we can't build each component by its own anymore, doesn't it?

**Comment by morantr** [comment]
I think we should wait before merge it until we will merge all the changes from the turris omnia.

**Comment by arnout** [comment]
> I think we should wait before merge it until we will merge all the changes from the turris omnia.

OK for me - it indeed creates conflicts that have to be resolved. Then I can include agent and controller into the top-level build in the same PR as well.

And anyway, CI is currently failing on this so it can't be merged as is.

File: CMakeLists.txt
Comment (vitalybu) [inline CMakeLists.txt]: Incorrect option description. Should be "Build EasyMesh controller".

File: CMakeLists.txt
Comment (vitalybu) [inline CMakeLists.txt]: Missing new line at the end of the file

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (vitalybu) [inline agent/src/beerocks/slave/son_slave_thread.cpp:126]: A better solution is to change the type of the controller socket to `SocketClient` in the `agent_ucc_listener` class.
Already did it in the UGW Support branch, so can be fixed later.

File: common/beerocks/btl/CMakeLists.txt
Comment (vitalybu) [inline common/beerocks/btl/CMakeLists.txt]: Why is this required anyway?
It makes sense for BWL and BPL as they are loaded dynamically, but for the BCL that it linked during build time, `target_link_libraries` should be enough, no?
If not, perhaps we should update the target to correctly set the INTERFACE_INCLUDE_DIRECTORIES.

File: common/beerocks/bwl/brcm/ap_wlan_hal_brcm.h
Comment (vitalybu) [inline common/beerocks/bwl/brcm/ap_wlan_hal_brcm.h]: I think that relative paths for files in the same module is better practice.
Can prevent issues when files from different modules have the same name.

This comment is relevant for all BPL/BWL files.

File: controller/src/beerocks/master/CMakeLists.txt
Comment (vitalybu) [inline controller/src/beerocks/master/CMakeLists.txt]: I guess it's not necessary after all...
Should be removed from all other CMake files as well.

**Review by vitalybu — CHANGES_REQUESTED** [review]

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (tomereli) [reply agent/src/beerocks/slave/son_slave_thread.cpp:126]: ok, so once you merge the UGW Support branch you can override it.

File: common/beerocks/btl/CMakeLists.txt
Comment (tomereli) [reply common/beerocks/btl/CMakeLists.txt]: You're right - it's not needed, and neither does the external (which no longer exists nor the MODULE_PATH

File: common/beerocks/bwl/brcm/ap_wlan_hal_brcm.h
Comment (tomereli) [reply common/beerocks/bwl/brcm/ap_wlan_hal_brcm.h]: The idea is to be able to easily move the BWL/BPL to the framework so we don't maintain 2 different common libraries. I also agree with you that using only the file name is problematic - so how about a compromise, adding a `bwl` directory prefix? for example, `#include <bwl/base_wlan_hal.h>`

Comment (tomereli) [reply common/beerocks/bwl/brcm/ap_wlan_hal_brcm.h]: If you agree with this approach, I will add this change in a separate commit right after this one.

Comment (vitalybu) [reply common/beerocks/bwl/brcm/ap_wlan_hal_brcm.h]: You can keep the relative paths and still move the modules between the projects.
Relative paths are used only for including files from the same logical group of files (e.g. inside the BWL module)

Comment (tomereli) [reply common/beerocks/bwl/brcm/ap_wlan_hal_brcm.h]: OK, so I'll remove this commit. @arnout if you still think this is required, we change it later.

Comment (tomereli) [reply common/beerocks/bwl/brcm/ap_wlan_hal_brcm.h]: On second thought we decided to keep Arnout's change but to use relative paths to `<bwl/...>` `<bpl/,,,>` `<btl/,,` etc. So I will change this commit by Arnout accordingly.

**Review by arnout — COMMENTED** [review]
I don't agree with making btlvf a static lib.

1. The point of #469 is *not* to make everything static. Rather, the point is to make it *possible* to build static.
2. Linking a static library into a dynamic library only works if the static library is built with `-fPIC`. We happen to do that, but it's quite hidden.
3. This is really bad for code size: the btlvf (which is very large) will be linked into both libbml and the controller executable as weak symbols. LTO is not going to save is there.
4. It's not really needed. Even if btlvf is built as a shared library, that doesn't mean it has to be exported. Since bml is linked dynamically with btlvf, the dynamic linker will resolve it. So it is only needed if bml is built as a static library - but then this doesn't solve it, because the bml static library will *not* include btlvf so we still need to link with btlvf explicitly.
5. btlvf is no different from bcl, elpp, tvlf, so if we do this, the same should be done for those libraries.

So, I removed the last two commits and pushed again. That's what I will merge.

I pushed the last two commits as branch feature/clean-up-bml-export.

I'll make a new branch that does further cmake cleanups which will include most of that, except for the shared->static change.

**Pull request merged**
